### PR TITLE
Use-actual-api_CU-m18rxy

### PR DESCRIPTION
**Pull request by a-wilkes**

### CODE REVIEW

**Comment by a-wilkes** [comment]
Task linked: [CU-m18rxy Use actual api](https://app.clickup.com/t/m18rxy)

**Comment by lpboucher** [comment]
Task linked: [CU-m18rxy Use actual api](https://app.clickup.com/t/m18rxy)

File: src/app/core/products/coffee-api.service.ts [hunk 1]
```diff
@@ -22,10 +22,7 @@ export class CoffeeApiService extends DataApiService<Coffee> {
     }
 
     getCoffee(slug: string): Observable<Coffee | undefined> {
```
Comment (lpboucher) [inline src/app/core/products/coffee-api.service.ts]: i think we can remove the undefined type

File: src/app/core/products/merchandise-api.service.ts [hunk 2]
```diff
@@ -1,36 +1,25 @@
...
+    constructor(protected http: HttpClient) {
+        super(config.backendURL + "wholesale/", http, Merchandise);
+    }
 
     getMerchandise(): Observable<Merchandise[]> {
-        return of(this.merchandise);
+        return this.getAll("merchandises?isActive=true");
     }
 
     getMerchandiseItem(slug: string): Observable<Merchandise | undefined> {
```
Comment (lpboucher) [inline src/app/core/products/merchandise-api.service.ts]: i think we can remove the undefined type

File: src/app/core/products/product-api.service.ts [hunk 3]
```diff
@@ -26,12 +26,8 @@ export class ProductApiService extends DataApiService<Product> {
...
 
     getProducts(): Observable<Product[]> {
-        return forkJoin([
-            this.coffeeApiService.getCoffees(),
-            this.merchandiseApiService.getMerchandise(),
-        ]).pipe(
-            map((products: Product[][]) => flatten(products)),
-        );
+        return zip(this.coffeeApiService.getCoffees(), this.merchandiseApiService.getMerchandise())
+            .pipe(map((products: Product[][]) => flatten(products)));
```
Comment (lpboucher) [inline src/app/core/products/product-api.service.ts:30]: i agree with you this is hard to understand. Let's revisit at a later point. I think we can one/a combination of rxjs operators that do a better job of conveying what is happening here

File: src/app/core/products/product-api.service.ts [hunk 4]
```diff
@@ -51,13 +47,20 @@ export class ProductApiService extends DataApiService<Product> {
     }
 
     getProductsByType(productType?: ProductType | "all"): Observable<Product[]> {
-        if (!productType || productType === "all") {
-            return this.getProducts();
+        switch (productType) {
+            case undefined: /* fallthrough */
```
Comment (lpboucher) [inline src/app/core/products/product-api.service.ts]: is this needed?

fallthrough will just fall to the default case no? Also, the default case in this example should be all

In any case, use a hashmap. Switch statements have O(n) complexity, while objects (with unique keys) are O(1)

I prefer) use an object (e.g dict) where the keys are the product types (all, coffee, etc.) the values are the returned observables.

And then simply return dict[productType] || dict["all"].

**Review by lpboucher — CHANGES_REQUESTED** [review]

**Review by lpboucher — APPROVED** [review]

**Pull request merged**